### PR TITLE
OKTA-576951 : Registration callbacks implementation for g3

### PR DESCRIPTION
**Pull request by glenfannin-okta**
## Description:

The purpose of this PR is to implement the Registration callbacks for the G3 SIW. 

** Note **
The assumption made here is that even though the API does not accept new (unregistered) fields to be passed successfully (without triggering an error), to still allow this in G3 because it currently works this way in G2. Customers _could_ be adding fields onto the page with the parseSchema callback, then, before submission to the API, using the preSubmit callback to collect the data and remove it from the payload before it is sent on to the API (avoid the IDX error for trying to utilize unregistered fields).

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [x] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [x] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-576951](https://oktainc.atlassian.net/browse/OKTA-576951)

### Screenshot/Video:
Registration callback demo:
https://drive.google.com/file/d/1bVrKfmrk4XheiqBpaewG-si-5fIFeqXK/view?usp=sharing

Added update to display monochrome style link within the ReminderPrompt component: 
image

image

### Downstream Monolith Build:

https://bacon-go.aue1e.saasure.net/commits?artifact=okta-signin-widget&page=1&pageSize=6&sha=d083d152f54199df30c59872f52f7b1c32e80c74&tab=main

### CODE REVIEW

File: src/types/options.ts [hunk 1]
```diff
@@ -137,7 +137,7 @@ interface Piv {
 }
 
 // Types for config
-type UserOperation =
+export type UserOperation =
```
Comment (glenfannin-okta) [inline src/types/options.ts:140]: Wanted to re-use this type without re-defining it in v3 folder.

File: src/v3/src/components/Form/renderers.tsx [hunk 2]
```diff
@@ -201,7 +202,7 @@ export default [
     renderer: InputPassword,
   },
   {
-    tester: ({ options: { inputMeta: { type } = {} as Input } }: FieldElement) => type === 'boolean',
+    tester: ({ options: { inputMeta: { type } = {} as Input } }: FieldElement) => type === 'boolean' || type === 'checkbox',
```
Comment (glenfannin-okta) [inline src/v3/src/components/Form/renderers.tsx:205]: Customers can send in `checkbox` as the `Input` type and it is expected to be respected and render as such (tested against v2).

File: src/v3/src/components/Form/renderers.tsx [hunk 3]
```diff
@@ -167,21 +167,22 @@ export default [
...
     tester: ({
       options: {
-        inputMeta: { options } = {} as Input,
+        inputMeta: { options, type } = {} as Input,
         format,
         customOptions,
       },
-    }: FieldElement) => (Array.isArray(customOptions) || Array.isArray(options)) && format === 'select',
+    }: FieldElement) => (Array.isArray(customOptions) || Array.isArray(options) || typeof options === 'object')
+      && ([format, type].includes('select')),
```
Comment (glenfannin-okta) [inline src/v3/src/components/Form/renderers.tsx:185]: Customers can send in `select` as the `Input` type and it is expected to be respected and render as such (tested against v2).

File: src/v3/src/components/Form/renderers.tsx [hunk 4]
```diff
@@ -167,21 +167,22 @@ export default [
   {
     tester: ({
       options: {
-        inputMeta: { options } = {} as Input,
+        inputMeta: { options, type } = {} as Input,
         format,
         customOptions,
       },
-    }: FieldElement) => (Array.isArray(customOptions) || Array.isArray(options)) && format === 'radio',
+    }: FieldElement) => (Array.isArray(customOptions) || Array.isArray(options)) && ([format, type].includes('radio')),
```
Comment (glenfannin-okta) [inline src/v3/src/components/Form/renderers.tsx:174]: Customers can send in `radio` as the `Input` type and it is expected to be respected and render as such (tested against v2).

File: src/v3/src/components/Select/Select.tsx [hunk 5]
```diff
@@ -47,6 +47,20 @@ const Select: UISchemaElementComponent<UISchemaElementComponentWithValidationPro
...
+  const getOptions = (): IdxOption[] | undefined => {
+    if (Array.isArray(customOptions) || Array.isArray(options)) {
+      return customOptions ?? options;
+    }
+
+    if (typeof options === 'object') {
+      return Object.entries(options)
+        .filter(([key, val]) => key !== '' && val !== '')
+        .map(([key, val]) => ({ label: val, value: key } as IdxOption));
+    }
```
Comment (glenfannin-okta) [inline src/v3/src/components/Select/Select.tsx:59]: Supports registration callback parseSchema format for select fields.

File: src/v3/src/hooks/useOnSubmit.ts [hunk 6]
```diff
@@ -140,36 +174,73 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
...
+          setMessage({
+            message: loc('oform.errorbanner.title', 'login'),
+            class: 'ERROR',
+            i18n: { key: 'oform.errorbanner.title' },
+          } as IdxMessage);
+        }
+        setStepToRender(stepToRender);
+        setLoading(false);
+        resolve?.(true);
+      };
```
Comment (glenfannin-okta) [inline src/v3/src/hooks/useOnSubmit.ts:202]: Needed a re-usable way to execute this group of functions (between the reg callback and in non registration flows/scenarios)

File: src/v3/src/transformer/field/type.ts [hunk 7]
```diff
@@ -39,6 +39,7 @@ const map = new Map<string, JsonSchemaType>([
   ['pdatetime', 'string'],
   ['ptime', 'string'],
   ['string', 'string'],
+  ['text', 'string'],
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/field/type.ts:42]: Registration parseSchema can send `text` as the `Input` type, so we need to make sure this is rendered properly.

File: src/v3/src/transformer/profile/transformEnrollProfile.ts [hunk 8]
```diff
@@ -97,6 +97,13 @@ export const transformEnrollProfile: IdxStepTransformer = ({ transaction, formBa
       validate: (data: FormBag['data']) => {
         const newPw = data['credentials.passcode'] as string;
         const errorMessages: WidgetMessage[] = [];
+        if (!newPw && passwordElement.options.inputMeta.required) {
+          errorMessages.push({
+            class: 'ERROR',
+            message: loc('model.validation.field.blank', 'login'),
+            i18n: { key: 'model.validation.field.blank' },
+          });
+        }
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/profile/transformEnrollProfile.ts:106]: Since we are overriding the default validation function, we need to check if the field is required and trigger the empty field error message.

File: src/v3/src/transformer/transaction/transform.ts [hunk 9]
```diff
@@ -44,6 +50,41 @@ const transformRemediationNameForIdp: TransformStepFnWithOptions = (options) =>
...
+    (error: APIError) => {
+      // Prioritize server side errors over custom error message
+      if (isClientTransaction) {
+        return;
+      }
+      triggerRegistrationErrorMessages(error, inputs, setMessage);
+    },
+  );
+  return formbag;
+};
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/transaction/transform.ts:85]: This is the meat of the parseSchema functionality.

TLDR;

I am taking the inputs sent to us from the API response and converting them into the Registration schema format (which is what v2 is currently expecting). Then, I am passing it to the callback (to allow the customer to add/update/etc) the schema, and in the body of the callback, I execute a function that converts the modified schema back into the IDX `Input` object array to allow it to pass through the SIW standard transformation functions (where we will get things like validation and element creation for free).

File: src/v3/src/types/schema.ts [hunk 10]
```diff
@@ -146,6 +146,15 @@ export type TokenReplacementValue = {
...
 export type TokenReplacement = Partial<Record<TokenSearchValue, TokenReplacementValue>>;
 
+export type RegistrationElementSchema = Modify<Input, {
+  'label-top'?: boolean;
+  placeholder?: string;
+  'data-se'?: string;
+  options?: IdxOption[] | Record<string, string>[];
+  sublabel?: string;
+  wide?: boolean;
+}>;
```
Comment (glenfannin-okta) [inline src/v3/src/types/schema.ts:156]: The v2 Registration schema is just a derivative of the `Input` object with some additional fields. 

I am going to file a new ticket to respect some of these new fields, i.e. `sublabel` should display a label under the main label, but it requires changes to many components and I wanted this PR to be as small as possible.

File: src/v3/src/types/widget.ts [hunk 11]
```diff
@@ -67,6 +75,21 @@ export type WidgetResetArgs = {
...
+    schema: RegistrationElementSchema[],
+    onSuccess: RegistrationSchemaCallbackV3,
+    onFailure: RegistrationErrorCallback
+  ) => void;
+  preSubmit?: (
+    postData: IdxActionParams,
+    onSuccess: RegistrationDataCallbackV3,
+    onFailure: RegistrationErrorCallback
+  ) => void;
+}>;
```
Comment (glenfannin-okta) [inline src/v3/src/types/widget.ts:91]: I had to modify this type because it looked like this was originally made for v1 (the current structure did not match v2 at all).

File: src/v3/src/util/idxUtils.ts [hunk 12]
```diff
@@ -229,3 +238,90 @@ export const convertIdxMessageToWidgetMessage = (
...
+  inputs.reduce((acc: Input[], input: Input) => {
+    const flattenedInputs = flattenInputs(input);
+    return [...acc, ...flattenedInputs];
+  }, []).map((flattendInput: Input) => ({
+    ...flattendInput,
+    'label-top': true,
+    'data-se': flattendInput.name,
+    wide: true,
+  } as RegistrationElementSchema))
+);
```
Comment (glenfannin-okta) [inline src/v3/src/util/idxUtils.ts:254]: The schema array that is sent to the customer in the callback is a flattened version of the `Input` array provided to use from IDX. So this function just flattens the objects and adds default values for those extra properties (which is what v2 sets them as too).

File: src/v3/src/util/idxUtils.ts [hunk 13]
```diff
@@ -229,3 +238,90 @@ export const convertIdxMessageToWidgetMessage = (
...
+        // Existing Field (modify)
+        Object.assign(idxInputs[existFieldIdx], schemaEle);
+      } else {
+        // New Field (create)
+        // @ts-expect-error Registration Schema Element has extra fields that are not in Input
+        idxInputs.push(schemaEle);
+      }
+    }
+  });
+};
```
Comment (glenfannin-okta) [inline src/v3/src/util/idxUtils.ts:302]: This function converts the flattened Registration schema back into an `Input` object array. There is an assumption made here that we wont nest objects more than 1 level (because testing in v2, those properties are ignore when I tried them).

File: src/v3/src/hooks/useHtmlContentParser.tsx [hunk 14]
```diff
@@ -23,7 +23,7 @@ import preact, { h } from 'preact';
 
 export const useHtmlContentParser = (
   content: string | undefined,
-  options?: HTMLReactParserOptions,
+  options?: HTMLReactParserOptions & { variant?: 'monochrome'; },
```
Comment (shawyu-okta) [inline src/v3/src/hooks/useHtmlContentParser.tsx:26]: This is a little bit smelly, we might want to think a little more (but not necessarily block this work) about how to do this without drilling the prop that doesn't really have anything superficially to do with this hook

Comment (glenfannin-okta) [reply src/v3/src/hooks/useHtmlContentParser.tsx:26]: I agree, initially what I thought I could do was just set the attribute on the link before we pass it to `loc`. What happens though is that the Parser doesn't recognize the `variant` attribute and strips it out :(

Comment (glenfannin-okta) [reply src/v3/src/hooks/useHtmlContentParser.tsx]: I found another way around this issue. I think this approach is a little cleaner and re-usable for other scenarios.
https://github.com/okta/okta-signin-widget/pull/3110/commits/d2c2ad7ba73fef0ee61392a819d90a54b614e4a2

File: src/v3/src/components/ReminderPrompt/ReminderPrompt.tsx [hunk 15]
```diff
@@ -119,6 +120,7 @@ const ReminderPrompt: UISchemaElementComponent<{
         <TextWithActionLink
           uischema={{
             type: 'TextWithActionLink',
+            parserOptions: uischema.parserOptions,
```
Comment (shawyu-okta) [inline src/v3/src/components/ReminderPrompt/ReminderPrompt.tsx:123]: Do we always want a monochrome link in `ReminderPrompt`? If so, we could hard-code that here instead of pulling from `uischema`

File: src/v3/src/transformer/transaction/transform.ts [hunk 16]
```diff
@@ -10,11 +10,17 @@
  * See the License for the specific language governing permissions and limitations under the License.
  */
 
-import { IdxRemediation } from '@okta/okta-auth-js';
-import { flow } from 'lodash';
```
Comment (shawyu-okta) [inline src/v3/src/transformer/transaction/transform.ts:14]: I _think_ this actually is the preferred way to import from `lodash` now: https://lodash.com/per-method-packages

**Review by shawyu-okta — APPROVED** [review]
strong work!

File: src/v3/src/util/settingsUtils.ts [hunk 17]
```diff
@@ -107,3 +120,88 @@ export const getDefaultCountryCode = (widgetProps: WidgetProps): string => {
...
+    (error: APIError) => {
+      const errorObj = error || {
+        errorSummary: loc('registration.default.callbackhook.error', 'login'),
+      };
+      onFailure(errorObj);
+    },
+  );
+};
+
+export const transformIdentifier = (
```
Comment (shawyu-okta) [inline src/v3/src/util/settingsUtils.ts:196]: Clarifying from yesterday's meeting, does this satisfy the `transformUsername` hook requirement?

Comment (glenfannin-okta) [reply src/v3/src/util/settingsUtils.ts:196]: Yep! It does! 🙏 I updated the doc to reference this.

File: src/v3/src/components/ReminderPrompt/ReminderPrompt.tsx (hunk 15 above)
Comment (glenfannin-okta) [reply src/v3/src/components/ReminderPrompt/ReminderPrompt.tsx:123]: Ah, so you mean call the function here to build the parser options (for the link replacer) instead of doing it at the transformer layer. I like it, it removes duplicated effort! 🙏
